### ISSUE TITLE
Improve performance of caching system and parsing

### ISSUE THREAD

**Issue by nevadascout** [HvyIndustries/crane#271]
## Improving writes on caching system :

Actually we flush all the cache into a single file. On big projects, that means every 5 sec after a change on a file, we have a flush over all the tree.

- 5 secs delay does not guarantee consistency of the tree in case of failure.
- By flushing all the tree, 5 secs may be a too short interval
- The tree may be large so I will use CPU & Disk intensively

We could fix this by :

- [ ] Separating ast symbols, one file file
- [ ] As flushing little files, maybe disabling zip support (no more need to use CPU for zip/unzip)

## Avoid parsing too early

Same trick as the cache flushing system, we could wait and trigger parsing every X seconds, instead at each document change, resulting in a decrease of CPU usage. Will be a good practice later for document validation.

**Comment by ichiriac** [comment]
BTW, it's already fully implemented in php-reflection (feature https://github.com/HvyIndustries/crane/issues/207) by using grafine engine :

- grafine uses a shards system
- nodes are distributed in shards by using [a round-robin algorithm](https://en.wikipedia.org/wiki/Round-robin_scheduling)
- distribution is sequenced in order to keep together related nodes
- grafine index & nodes (shards contents) are retrieved with a lazy loading system

Resulting in separate files, each weighted proportionnaly to symbols amounts, so a little 1 symbol file will not result into a small json file, the system makes so the I/O disk + JSON parsing as small as possible (too many small files will decrease speeds).

Futhermore when you start loading a huge 20k classes project / with > 50k variables declarations/usage, at startup, the loading time/memory usage will be as small as an empty project as it will load data in-memory on demand.

Symbols like classes, interfaces, traits, functions, namespaces are indexed, so access times on a specific class is under 1ms (no mater the size of data). 

> But eager search functions like get a list of all classes starting by `foo%` will trigger full index loads (not nodes, they allways loads in a lazy fashion)

PHP-Reflection at full scanning / re-index time adds a cache layer based on filesize/mtime/md5 in order to avoid parsing already updated files.

I'm working on components communications, because the mais problem here are CPU intensive tasks, and lack of threads in node. THe final structure (process oriented) will be the following :

Your Process -> running an object instance of php-reflection
P1 -> process of grafine (running indexes / requests / data I/O)
P2 -> worker parsing a file and sending nodes to P1/grafine
... (n=number of physical processors)
PN -> worker

The parsing a file workflow :

```
Your Server Process -> php-reflection instance 
  -> call a parsing function 
    -> 1. check caching (if not changed returns without parsing)
    -> 2. sends the request to a load balancer of workers
       -> 1. Worker : unstack the last parsing request and do its job
       -> 2. Worker : at end analyse AST and converts in nodes
       -> 3. Worker : sends nodes to grafine
           -> 1. Grafine : release old nodes if already existing
           -> 2. Appends nodes / indexes (lazy load them if not already loaded)
```

> As workers are amnesic (grafine is the only one who knows all data) we can kill them without any problem. A timeout of 5 sec and a restart in case of faillure can now be implemented (https://github.com/glayzzle/php-reflection/blob/master/src/worker.js#L76)

Also grafine have a native support of sharding, and the index management is separated,, so it will be room to make optimisations for very huge projects (millions of line of code / if RAM memory is available)  : a process for searching, and multiple process for storing and manipulating nodes / shards. This will enable to search between symbols at an optimal speed during intensively creating node. 

All of these comes with the IPC overhead price, so microbench will be a bit worse, but the system is scallable (no limit). The only advantage will be the multi processor support that will overcome the IPC overhead by a bit, but as the parsing is made at high speed and the most cost effective is the disk I/O, it will be true only for huge/complexe php files.

---

As performances are not too bad in Crane for now, maybe this task may be replaced with the migration on php-reflection

**Comment by nevadascout** [comment]
Closing this as it will be implemented properly by #207
